[READY] Use SetQuickFixList function for GoTo* subcommands

micbou opened the pull request:
From discussion in PR #2101, we reuse the `SetQuickFixList` function for the `GoTo*` subcommands. This automatically makes the quickfix window height equal to the number of locations.

Unfortunately, there is no Vim option to change the default height (10) of the quickfix window so we can't really use the heuristic `min(number of Locations, default height)` because users would not be able to change the default height (and we don't want to add another option). However, there are still solutions to set a default height. For instance, a user could add the following line in its vimrc:
```viml
autocmd FileType qf exe "20wincmd _"
```
to set the default height to 20. This will override YCM behavior.

I didn't reproduce the `lazyredraw` boilerplate from `youcompleteme#OpenGoToList`. I guess it was used to avoid flickering but I didn't experienced any without it. In addition, it was not properly implemented because the `lazyredraw` state was not preserved if the option was already set.
I also threw away the `cclose` command. It resets the quickfix window to its default position if already opened, which I think is not a desirable behavior.

Closes #2101.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2190)
<!-- Reviewable:end -->

Valloric commented:
Thanks for the PR! I like this quite a bit. There are some issues to resolve though; check inline comments.

> autocmd FileType qf exe "20wincmd _"

We might want to add the above to the FAQ.



---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 828 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8MuKFBCQHSjuGCrSp:-KJ8MuKFBCQHSjuGCrSq:314366743) ([raw file](https://github.com/valloric/youcompleteme/blob/8b1fba22cc50c2fd004881ebc6fc64246fe84592/autoload/youcompleteme.vim#L828)):*
> ```VimL
> 
> 
> function! youcompleteme#OpenGoToList()
> ```

Unfortunately this is a public function so we can't just delete it without breaking backwards compatibility (which we don't want to do). Can we forward this to the python function?

---

*[python/ycm/vimsupport.py, line 266 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8NDK2GK_cJmZdXKy3:-KJ8NDK2GK_cJmZdXKy4:-581506438) ([raw file](https://github.com/valloric/youcompleteme/blob/ed45003f382cc7ad32f348373966d39035e5882c/python/ycm/vimsupport.py#L266)):*
> ```Python
>     JumpToPreviousWindow()
>   if autoclose:
>     vim.command( 'au WinLeave <buffer> q' )
> ```

Isn't this going to re-add the event handler every time?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-:-KJ8NS4M_UfyGZ-lrk48:960779243)*
<!-- Sent from Reviewable.io -->

micbou commented:
I improved how the height of the quickfix window is computed: it can now handle entries that span on multiple lines. With this change, the `FileType` autocommand cannot be used anymore to change the default height (because the height is modified after opening the quickfix window) so I created the `YcmQuickFixOpened` autocommand in order to customize the way YCM is opening the quickfix window. I updated the README with this autocommand by creating a new section.

<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Thanks for the PR! I like this quite a bit. There are some issues to resolve though; check inline comments.
> 
> > autocmd FileType qf exe "20wincmd _"
> 
> We might want to add the above to the FAQ.
</details>


---

Reviewed 1 of 5 files at r1, 5 of 6 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 828 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8MuKFBCQHSjuGCrSp:-KJCKLL7EHZl8NrTUY35:1837642880) ([raw file](https://github.com/valloric/youcompleteme/blob/8b1fba22cc50c2fd004881ebc6fc64246fe84592/autoload/youcompleteme.vim#L828)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Unfortunately this is a public function so we can't just delete it without breaking backwards compatibility (which we don't want to do). Can we forward this to the python function?
</details>
This function now uses the new `OpenQuickFixList` function from `vimsupport`. I added a deprecation notice because I don't think this function should be public and this allows us to remove it later. Tell me if you feel it's not appropriate.

---

*[python/ycm/vimsupport.py, line 266 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8NDK2GK_cJmZdXKy3:-KJAedit7XhZeMj1OWmK:1809108039) ([raw file](https://github.com/valloric/youcompleteme/blob/ed45003f382cc7ad32f348373966d39035e5882c/python/ycm/vimsupport.py#L266)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Isn't this going to re-add the event handler every time?
</details>
I checked with `:set verbose=6` and the autocommand is removed when the quickfix window is closed so we should be fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-:-KJCM51_jKsohrKTSuTi:1641284783)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, micbou wrote…</i></summary>
> I improved how the height of the quickfix window is computed: it can now handle entries that span on multiple lines. With this change, the `FileType` autocommand cannot be used anymore to change the default height (because the height is modified after opening the quickfix window) so I created the `YcmQuickFixOpened` autocommand in order to customize the way YCM is opening the quickfix window. I updated the README with this autocommand by creating a new section.
</details>


---

Reviewed 1 of 5 files at r1, 1 of 6 files at r2, 2 of 2 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-:-KJHlY196zzZIbvB9Rmg:1417088721)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 1517 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJIYrqJ0GQWFk3ApdR-:-KJIYrqJ0GQWFk3ApdR0:-834193655) ([raw file](https://github.com/valloric/youcompleteme/blob/5e35e60fb0ac71dda034c40240f976d1965c17b6/README.md#L1517)):*
> ```Markdown
> endfunction
> 
> autocmd User YcmQuickFixOpened call s:CustomizeYcmQuickFixWindow()
> ```

Huh, never heard of a User autocommand! Is this new? If yes, is it supported by the min version of Vim we support?

---

*[autoload/youcompleteme.vim, line 828 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8MuKFBCQHSjuGCrSp:-KJIZ3pRchroi8AVhxAk:-2048659454) ([raw file](https://github.com/valloric/youcompleteme/blob/8b1fba22cc50c2fd004881ebc6fc64246fe84592/autoload/youcompleteme.vim#L828)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> This function now uses the new `OpenQuickFixList` function from `vimsupport`. I added a deprecation notice because I don't think this function should be public and this allows us to remove it later. Tell me if you feel it's not appropriate.
</details>
I'm not at all against leaving a deprecation notice, and you might even be right in that we should have never exposed it publicly (I don't remember the rationale).

But I don't think the user particularly cares about _why_ this is being deprecated, so I'd remove the sentence about "this should not have been exposed." One of the sad laws of UX design is that the longer the error message, the more likely it is that the user won't read it. 

---

*[python/ycm/vimsupport.py, line 266 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8NDK2GK_cJmZdXKy3:-KJIZecnaodkNMRWuvea:405159664) ([raw file](https://github.com/valloric/youcompleteme/blob/ed45003f382cc7ad32f348373966d39035e5882c/python/ycm/vimsupport.py#L266)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> I checked with `:set verbose=6` and the autocommand is removed when the quickfix window is closed so we should be fine.
</details>
Awesome! Could you put a comment here stating that for future readers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 1517 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJIYrqJ0GQWFk3ApdR-:-KJIa_bVAnHDDIDe_Jrg:-1221202877) ([raw file](https://github.com/valloric/youcompleteme/blob/5e35e60fb0ac71dda034c40240f976d1965c17b6/README.md#L1517)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Huh, never heard of a User autocommand! Is this new? If yes, is it supported by the min version of Vim we support?
</details>
No, they're not new.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 3 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 1517 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJIYrqJ0GQWFk3ApdR-:-KJIge7okUHaxaDp9Jxw:1084709044) ([raw file](https://github.com/valloric/youcompleteme/blob/5e35e60fb0ac71dda034c40240f976d1965c17b6/README.md#L1517)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> No, they're not new.
</details>
It was introduced in Vim 5.0. I think we are good.

---

*[autoload/youcompleteme.vim, line 828 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8MuKFBCQHSjuGCrSp:-KJIly6uepnM-ZdWJxqW:-51467578) ([raw file](https://github.com/valloric/youcompleteme/blob/8b1fba22cc50c2fd004881ebc6fc64246fe84592/autoload/youcompleteme.vim#L828)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I'm not at all against leaving a deprecation notice, and you might even be right in that we should have never exposed it publicly (I don't remember the rationale).
> 
> But I don't think the user particularly cares about _why_ this is being deprecated, so I'd remove the sentence about "this should not have been exposed." One of the sad laws of UX design is that the longer the error message, the more likely it is that the user won't read it. 
</details>
I removed the `This function was never intended to be exposed.` part.

---

*[python/ycm/vimsupport.py, line 266 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-KJ8NDK2GK_cJmZdXKy3:-KJIlqrzrvCn_Oj_wIDU:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/ed45003f382cc7ad32f348373966d39035e5882c/python/ycm/vimsupport.py#L266)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Awesome! Could you put a comment here stating that for future readers?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #2204) made this pull request unmergeable. Please resolve the merge conflicts.

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from me. Would trigger homu but this has merge conflicts.

<details><summary><i>Previously, homu (Homu) wrote…</i></summary>
> :umbrella: The latest upstream changes (presumably #2204) made this pull request unmergeable. Please resolve the merge conflicts.
</details>


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-:-KK5MZMJpQjph47PLCgD:246089777)*
<!-- Sent from Reviewable.io -->

micbou commented:
Fixed the conflicts.

When I have time, I'll send a PR to add an autocommand similar to `YcmQuickFixOpened` but for the location list.

@homu r=Valloric

<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> :lgtm: from me. Would trigger homu but this has merge conflicts.
</details>


---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2190#-:-KK5mBX4ZZHZiyh0qkp7:1675736895)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 805911b has been approved by `Valloric`

<!-- @homu r=Valloric 805911b56b9ce204af0552994df45c83e1674bbd -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/137129308)
